compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp

```diff
@@ -690,6 +690,7 @@ static void addLowerToLLVMPasses(OpPassManager &passManager) {
 
 void buildLLVMCPUCodegenPassPipeline(OpPassManager &passManager) {
   addCommonTargetExecutablePreprocessingPasses(passManager.nest<ModuleOp>());
+  passManager.addPass(createExpandF16MaxFToF32Pass());
```

Comment:
Is this where it always was? I thought this was supposed to be done in the `addLowerToLLVMPasses` call.

Reply:
I think it can be added before ArithExpandOpsPass.

https://github.com/openxla/iree/blob/8c41e5177f6a8cc33200bb91c8b01dee72bf0942/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L673-L675

Reply:
It seems to give me the same fmaxf error, if I add the pass in addLowerToLLVMPasses.
I tried adding it before passManager.addNestedPass<func::FuncOp>(arith::createArithExpandOpsPass()); 
But run into error. I also tried adding it even before in addLowerToLLVMPasses, but I still get the fmaxf error

Reply:
What is the error? Can you share the input MLIR file and the IR dumps?

Reply:
Could you post the output of `iree-compile` with `--mlir-disable-threading --mlir-print-ir-after-all --mlir-print-ir-before-all` somewhere. It is going to dump a lot of IR so do 
```
iree-compiler <other-flags> --mlir-disable-threading --mlir-print-ir-after-all --mlir-print-ir-before-all <input_file> -o test.vmfb 2> debug.log
```

and upload `debug.log` somewhere (easiest is on `gist.github.com`)

Reply:
[linalg_mlir.txt](https://github.com/openxla/iree/files/11617391/linalg_mlir.txt)
[linalg_error.txt](https://github.com/openxla/iree/files/11617409/linalg_error.txt)


This is the error I get 

lld: error: undefined symbol: fmaxf
>>> referenced by -:4
>>>               /tmp/forward_dispatch_42_softmax_2x5x4096x4096xf16_dispatch_0-f3c0df.o:(forward_dispatch_42_softmax_2x5x4096x4096xf16_dispatch_0_softmax_2x5x4096x4096xf16)
>>> referenced by -:4
>>>               /tmp/forward_dispatch_42_softmax_2x5x4096x4096xf16_dispatch_0-f3c0df.o:(forward_dispatch_42_softmax_2x5x4096x4096xf16_dispatch_0_softmax_2x5x4096x4096xf16)
>>> referenced by -:4
>>>               /tmp/forward_dispatch_42_softmax_2x5x4096x4096xf16_dispatch_0-f3c0df.o:(forward_dispatch_42_softmax_2x5x4096x4096xf16_dispatch_0_softmax_2x5x4096x4096xf16)

Reply:
> In response, we decided to write the pass into its own file: ExpandF16MaxFToF32Pass.cpp. We added this pass to Passes.cpp and LLVMCPUPasses.h. This alternative approach appears to resolve the issue, enabling successful 16bit MLIR compilation.

I thought that you put it to the beginning because it would work. But it seems not. The log you attached is using the current order, and it produces the error? Do I misunderstand something?

side question -- I mis-track if we are able to load/store fp16 in CPU backend. Will we need emulation for fp16 load/store?

Reply:
Hey @hanhanW, @Ranvirsv has tried adding it right above `ArithExpandOpsPass` but it was failing, so he tried again putting it a bit earlier, i.e at the start of `addLowerToLLVMPasses`. However, both of these places ends up generating the error which he uploaded above.

The only place that has worked so far(**compiles + without error**) is in `buildLLVMCPUCodegenPassPipeline`.

Reply:
Whats happening here is that the `fmaxf` is being generated by the LLVM backend... I still cant really see the error. What is atteched as `linalg_error.mlir` is with the pass being run early vs being run late..... Really it shouldnt matter whether it is run there or before `addLowerToLLVMPasses`. So looking at the IR when the pass is added late would help. Maybe there is a reason why this expansion isnt triggering when added late.

Reply:
I am sorry, this is the file with the error log
This is when running the pass late (in ```addLowerToLLVMPasses```) 

[Uploading linalg_error.log…]()

Reply:
Ok there are a couple of issues here....
1) I think your pass is not geared to handle vector types, and at this stage after this is vectorized you get a `vector.reduction` of `maxf` 
```
 %51 = vector.reduction <maxf>, %28, %50 : vector<32xf16> into f16
```
So doing this all the way at the start is basically avoiding this, so its kind of a WAR.....
@raikonenfnu how important is this change for you guys? Its a legit workaround... but it is hiding the deeper issue. So trying to balance these.

2) @dcaballe and @hanhanW this is another instance of issue we have seen in different places.... There is a lot of `vector` ops being generated. We need to handle this..... Is there already an issue tracking this. We probably need an EPIC that can capture all bugs related to this issue.

Reply:
This is required for LLaMA / Vicuna FP16 models on CPU. Very high priority for us so we can at least do FP16 for LLMs on CPU with IREE (llama.cpp etc are on int4 and we have to get to that soon).

Reply:
Sigh... Well we need to fix the overall vectorization story . But to unblock for now adding it here is fine. Could you please add a TODO to say this pass is being done way too late.... and it needs to happen during lowering to LLVM. Could you also file an issue documenting the reason why this is being done here.

Reply:
Not sure what I follow the problem here. Would you mind elaborating? What do you mean by "There is a lot of vector ops being generated"? 

If this is expanding f16 -> f32 max ops, this should happen before vectorization and then the vectorizer should generate the corresponding f32 max reduction. After that, the maxf operation should be expanded to a cmp/select op (I'm assuming that the frontend has generated all the handling for the fp corner cases (ie., NaN, inf, etc.)).

Reply:
Sure sound's good.

Reply:
I'm surprised that it works if you only convert maxf op. Why only maxf op is needed?

@MaheshRavishankar , the reduction on f16 is generated because other operations are not expanded to f32.

```
 %51 = vector.reduction <maxf>, %28, %50 : vector<32xf16> into f16
```

Shouldn't we also handle other arith ops like subf, extf, addf?

https://gist.githubusercontent.com/hanhanW/1f85a07d1d3e75309168fc4d01f8be17/raw/ae9e627246b444c94c0fa49327a0329034b4104d/a.mlir

---

The other issue is that the first generic op is not vectorized even it has a static shape. Maybe should fie an issue and assign it to Diego..

https://gist.github.com/hanhanW/d3ba1367ed79236217ffc4c24f0db319

Reply:
RE @dcaballe about "There is a lot of vector ops being generated", I think it means that we don't tile the operations enough. In this case, we unroll some huge vectors, which results in a lot of vector ops before LLVM conversion.

This one is an other example that we need multi lowering config, though it is more complicated. It is reduction -> reduction -> element with triangle deps..

Reply:
> I'm surprised that it works if you only convert maxf op. Why only maxf op is needed?
> 
> @MaheshRavishankar , the reduction on f16 is generated because other operations are not expanded to f32.
> 
> ```
>  %51 = vector.reduction <maxf>, %28, %50 : vector<32xf16> into f16
> ```
> 

THe issue is `fmaxf` lowers to a libc function while others dont....

Reply:
@dcaballe see #13888 . Lets carry over the thread there and let @Ranvirsv land this PR after all other minor comments are addressed.

Reply:
> Sigh... Well we need to fix the overall vectorization story . But to unblock for now adding it here is fine. Could you please add a TODO to say this pass is being done way too late.... and it needs to happen during lowering to LLVM. Could you also file an issue documenting the reason why this is being done here. So file and issue and add a
> 
> ```
> // TODO(#<issue_number) : <reason>
> ```

is this the issue ``` Vector lowering results in IR explosion #13888 ```, or should I make a different issue?

Reply:
You can use #13888 (also @hanhanW I filed an issue with the log)